ml/Dimension.cc

```diff
@@ -144,30 +266,80 @@ bool Dimension::predict(CalculatedNumber Value, bool Exists) {
     const DSample Sample = SB.preprocess().back();
     delete[] TmpCNs;
 
+    /*
+     * Lock to predict and possibly schedule the dimension for training
+    */
+
     std::unique_lock<std::mutex> Lock(Mutex, std::defer_lock);
     if (!Lock.try_lock()) {
-        AnomalyBit = false;
         return false;
     }
 
+    // Mark the metric time as variable if we received different values
+    if (!SameValue)
+        MT = MetricType::Variable;
```

Comment:
Hello @vkalintiris ,

I am on vacation, so I am only giving to you this coverity scan report:

```sh
** CID 432006:  Concurrent data access violations  (MISSING_LOCK)
/ml/Dimension.cc: 280 in ml::Dimension::predict(long, double, bool)()
274         if (!Lock.try_lock()) {
275             return false;
276         }
277     
278         // Mark the metric time as variable if we received different values
279         if (!SameValue)
>>>     CID 432006:  Concurrent data access violations  (MISSING_LOCK)
>>>     Accessing "this->MT" without holding lock "ml::Dimension.Mutex". Elsewhere, "ml::Dimension.MT" is accessed with "Dimension.Mutex" held 2 out of 3 times (1 of these accesses strongly imply that it is necessary).
280             MT = MetricType::Variable;
281     
282         // Decide if the dimension needs to be scheduled for training
283         scheduleForTraining(CurrT);
284     
285         // Nothing to do if we don't have a model
```

The PR is running as expected, but I did not have time today to review the code.

Best regards!

Reply:
Don't think this specific report is accurate. We reach LOC 280 only if we managed to hold the lock earlier at LOC 274.

Reply:
When I reported initially I did not review code, now that I am reviewing, I agree with explanation so I am marking as resolved.